server/persistence/iwclient/iwclient_list.go

```diff
@@ -30,34 +31,48 @@ func (c *Client) ListAsUser(ctx context.Context, user string, workspaces *worksp
 }
 
 func (c *Client) fetchMissingWorkspaces(ctx context.Context, user string, workspaces *workspacesv1alpha1.InternalWorkspaceList) error {
-	// list user's space bindings
-	sbb := toolchainv1alpha1.SpaceBindingList{}
-	if err := c.listUserSpaceBindings(ctx, user, &sbb); err != nil {
+	// fetch all workspaces
+	aww := workspacesv1alpha1.InternalWorkspaceList{}
+	if err := c.backend.List(ctx, &aww); err != nil {
+		return err
+	}
+
+	// retrieve names of missing workspaces
+	nmww, err := c.calculateNamesOfMissingWorkspaces(ctx, user, workspaces)
+	if err != nil {
 		return err
 	}
 
-	// filter already fetched Workspaces
-	fsbb := make([]*toolchainv1alpha1.SpaceBinding, 0, len(sbb.Items))
-	for i, sb := range sbb.Items {
-		if slices.ContainsFunc(workspaces.Items, func(w workspacesv1alpha1.InternalWorkspace) bool {
-			return w.Name == sb.Spec.Space
-		}) {
-			continue
+	// add workspaces to which the user has direct access to return list
+	for _, s := range nmww {
+		// TODO(@filariow): use a field selector after #195 lands
+		if i := slices.IndexFunc(aww.Items, func(w workspacesv1alpha1.InternalWorkspace) bool {
+			return w.Status.Space.Name == s
+		}); i != -1 {
+			workspaces.Items = append(workspaces.Items, aww.Items[i])
 		}
+	}
+	return nil
+}
 
-		fsbb = append(fsbb, &sbb.Items[i])
+func (c *Client) calculateNamesOfMissingWorkspaces(ctx context.Context, user string, workspaces *workspacesv1alpha1.InternalWorkspaceList) ([]string, error) {
+	// list user's space bindings
+	sbb := toolchainv1alpha1.SpaceBindingList{}
+	if err := c.listUserSpaceBindings(ctx, user, &sbb); err != nil {
+		return nil, err
 	}
 
-	for _, sb := range fsbb {
-		k := c.workspaceNamespacedName(sb.Spec.Space)
-		w := workspacesv1alpha1.InternalWorkspace{}
-		if err := c.backend.Get(ctx, k, &w, &client.GetOptions{}); err != nil {
-			continue
-		}
+	wwn := sets.New[string]()
+	for _, w := range workspaces.Items {
+		wwn = wwn.Insert(w.Status.Space.Name)
+	}
 
-		workspaces.Items = append(workspaces.Items, w)
+	sbbn := sets.New[string]()
+	for _, sb := range sbb.Items {
+		sbbn.Insert(sb.Spec.Space)
```

Comment:
`sets.Insert` doesn't mutate, so you need to assign its output back to the set.
```suggestion
		sbbn = sbbn.Insert(sb.Spec.Space)
```

Alternatively, would it be more efficient to remove the elements from `wwn` and skip making `sbbn`?

Reply:
great catch, let me fix that!

Reply:
done in 0eed33e